File: src/main/java/ru/nucodelabs/gem/app/snapshot/HistoryManager.java
package ru.nucodelabs.gem.app.snapshot;


import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

import static java.lang.Math.max;
import static java.lang.Math.min;

public class HistoryManager<T> {

    private final Snapshot.Originator<T> originator;
    private List<Snapshot<T>> history = new ArrayList<>();
    private int position = 0;

    @Inject
    public HistoryManager(Snapshot.Originator<T> originator) {
        this.originator = originator;
    }

    public void snapshot() {
        Snapshot<T> snapshot = originator.getSnapshot();
        if (history.isEmpty() || !Objects.equals(history.get(position), snapshot)) {
            if (position < history.size() - 1) {
                history = history.subList(0, max(0, position + 1));
            }
            history.add(snapshot);
            position = history.size() - 1;
        }
    }

    public void performThenSnapshot(Runnable operation) {
        operation.run();
        snapshot();
    }

    public void redo() {
        getRedo().ifPresent(originator::restoreFromSnapshot);
    }


    public void undo() {
        getUndo().ifPresent(originator::restoreFromSnapshot);
    }

    private Optional<Snapshot<T>> getUndo() {
        if (position == 0 || history.isEmpty()) {
            return Optional.empty();
        }
        position = max(0, position - 1);
        return Optional.of(history.get(position));
    }

    private Optional<Snapshot<T>> getRedo() {
        if (position == history.size() - 1 || history.isEmpty()) {
            return Optional.empty();
        }
        position = min(history.size() - 1, position + 1);
        return Optional.of(history.get(position));
    }

    public void clear() {
        history.clear();
        position = 0;
    }
}

File: src/main/java/ru/nucodelabs/gem/app/snapshot/HistoryManager.kt
package ru.nucodelabs.gem.app.snapshot

import ru.nucodelabs.gem.app.snapshot.Snapshot.Originator
import javax.inject.Inject

class HistoryManager<T> @Inject constructor(private val originator: Originator<T>) {
    private var history: MutableList<Snapshot<T>> = mutableListOf()
    private var position = 0

    /**
     * Makes a snapshot of originator
     */
    fun snapshot() {
        val snapshot = originator.snapshot()
        if (history.isEmpty() || history[position] != snapshot) {
            if (position < history.lastIndex) {
                history = history.subList(0, position + 1)
            }
            history += snapshot
            position = history.lastIndex
        }
    }

    /**
     * Makes a snapshot after executing block
     */
    fun snapshotAfter(block: () -> Unit) {
        block()
        snapshot()
    }

    /**
     * Restores originator from next snapshot in history if it exists else do nothing
     */
    fun redo() = snapshotToRedo()?.let { originator.restoreFromSnapshot(it) }

    /**
     * Restore originator from previous snapshot in history if it exists else do nothing
     */
    fun undo() = snapshotToUndo()?.let { originator.restoreFromSnapshot(it) }

    /**
     * Returns previous snapshot if it exists and change position
     */
    private fun snapshotToUndo(): Snapshot<T>? {
        if (history.isEmpty() || position == 0) {
            return null
        }
        position--
        return history[position]
    }

    /**
     * Returns next snapshot if it exists and changes position
     */
    private fun snapshotToRedo(): Snapshot<T>? {
        if (history.isEmpty() || position == history.lastIndex) {
            return null
        }
        position++
        return history[position]
    }

    /**
     * Reset
     */
    fun clear() {
        history.clear()
        position = 0
    }
}